lib/scheduler/thread_pool.rb

```diff
@@ -45,37 +50,45 @@ def post(&block)
 
       @mutex.synchronize do
         @queue << wrapped_block
-
-        spawn_thread if @queue.length > 1 && @threads.length < @max_threads
+        spawn_thread if @threads.length == 0
 
         @new_work.signal
       end
     end
 
-    def shutdown(timeout: 30)
-      @mutex.synchronize do
-        return if @shutdown
-        @shutdown = true
-        @threads.length.times { @queue << :shutdown }
-        @new_work.broadcast
-      end
-
+    def wait_for_termination(timeout: nil)
       threads_to_join = nil
       @mutex.synchronize { threads_to_join = @threads.to_a }
 
-      failed_to_shutdown = false
+      if timeout.nil?
+        threads_to_join.each(&:join)
+      else
+        failed_to_shutdown = false
+
+        deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
+        threads_to_join.each do |thread|
+          remaining_time = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
+          break if remaining_time <= 0
+          if !thread.join(remaining_time)
+            Rails.logger.error "ThreadPool: Failed to join thread within timeout\n#{thread.backtrace.join("\n")}"
+            failed_to_shutdown = true
+          end
+        end
 
-      deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
-      threads_to_join.each do |thread|
-        remaining_time = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
-        break if remaining_time <= 0
-        if !thread.join(remaining_time)
-          Rails.logger.error "ThreadPool: Failed to join thread within timeout\n#{thread.backtrace.join("\n")}"
-          failed_to_shutdown = true
+        if failed_to_shutdown
+          @mutex.synchronize { @threads.each(&:kill) }
+          raise ShutdownError, "Failed to shutdown ThreadPool within timeout" if failed_to_shutdown
```

Comment:
```suggestion
          raise ShutdownError, "Failed to shutdown ThreadPool within timeout"
```

This is already within the `if failed_to_shutdown` conditional.

Reply:
oops thanks